File: swift/Sources/QRlessPay/Uwb.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) OpenBank contributors.
import Foundation

/// Enhanced proximity (wire spec §5) — optional, never required.
///
/// **The spec describes UWB in prose and gives it no wire format.** It calls the capability
/// "negotiated, best-effort" without defining how a device announces it or how ranging parameters
/// cross between two phones. This is therefore a **v1.1 proposal**, mirroring `Uwb.kt`, and marked
/// as such wherever it surfaces.
///
/// Two facts decide the design, both unwelcome: **Apple's Nearby Interaction and Android's FiRa
/// stack do not interoperate**, so a cross-platform pair cannot range at any effort; and **UWB
/// hardware is a minority**, so RSSI stays the baseline and UWB may only sharpen it. Nothing here
/// may become a precondition for paying.
public enum Uwb {
    /// Advert flag proposed for "this device can range". Reserved-unused until the spec adopts it.
    public static let flagUwbCapable: UInt8 = 0x8
    /// Proposed fourth characteristic, for exchanging the token below.
    public static let tokenCharUuid = "0000C3A6-2F3B-4E8A-9A5E-0B0E6F1C2D3A"
}

/// A ranging token as it crosses the air.
///
/// The platforms disagree about what a token is. FiRa needs concrete session parameters and splits
/// **controller** (owns them) from **controlee** (joins); Apple's `NIDiscoveryToken` is an opaque
/// blob with no such split. All three shapes are carried and, critically, tagged — so an Apple
/// token arriving at a FiRa parser is *rejected* rather than read as a malformed FiRa token.
///
/// That rejection is the point: silently misparsing a foreign token yields a session reporting
/// distances derived from nonsense, which is worse than no ranging — it would show a payer a
/// confident "0.3 m" for a peer that could be anywhere.
public enum UwbToken: Equatable {
    case controller(address: [UInt8], channel: Int, preambleIndex: Int, sessionId: Int)
    case controlee(address: [UInt8])
    /// Apple `NIDiscoveryToken`, opaque by design — never parsed, only handed back to the platform.
    case opaque(blob: [UInt8])
}

/// `[magic 'Q'][version][kind][payload]`.
///
/// The magic byte and version exist so a token from the other platform, or a future revision, fails
/// to decode rather than decoding into something wrong.
public enum UwbTokenCodec {

    private static let magic = UInt8(ascii: "Q")
    private static let version: UInt8 = 1
    private static let kindController: UInt8 = 1
    private static let kindControlee: UInt8 = 2
    private static let kindOpaque: UInt8 = 3
    private static let header = 3
    private static let addressBytes = 2
    private static let controllerPayload = 2 + 1 + 1 + 4

    public static func encode(_ token: UwbToken) -> [UInt8] {
        switch token {
        case let .controller(address, channel, preambleIndex, sessionId):
            var out: [UInt8] = [magic, version, kindController]
            out += Array(address.prefix(addressBytes))
            out += [UInt8(channel & 0xFF), UInt8(preambleIndex & 0xFF)]
            out += (0..<4).reversed().map { UInt8((sessionId >> (8 * $0)) & 0xFF) }
            return out
        case let .controlee(address):
            return [magic, version, kindControlee] + Array(address.prefix(addressBytes))
        case let .opaque(blob):
            return [magic, version, kindOpaque] + blob
        }
    }

    /// Nil for anything that is not a token of this version — including the other platform's.
    public static func decode(_ bytes: [UInt8]) -> UwbToken? {
        guard bytes.count >= header, bytes[0] == magic, bytes[1] == version else { return nil }
        let payload = Array(bytes[header...])
        switch bytes[2] {
        case kindController:
            guard payload.count == controllerPayload else { return nil }
            let sessionId = payload[4...7].reduce(0) { ($0 << 8) | Int($1) }
            return .controller(
                address: Array(payload[0..<addressBytes]),
                channel: Int(payload[addressBytes]),
                preambleIndex: Int(payload[addressBytes + 1]),
                sessionId: sessionId
            )
        case kindControlee:
            guard payload.count == addressBytes else { return nil }
            return .controlee(address: payload)
        case kindOpaque:
            guard !payload.isEmpty else { return nil }
            return .opaque(blob: payload)
        default:
            return nil
        }
    }
}

/// What a ranging attempt concluded.
public enum ProximityOutcome: Equatable {
    /// UWB ranged and the peer is within `metres`.
    case ranged(metres: Double)
    /// No UWB session was possible. **Not an error** — the RSSI baseline already gated this tile,
    /// and the §6 confirmation authorises the payment either way.
    case downgraded(why: String)
}

/// Decides whether a ranging attempt is worth starting, before any radio is touched.
///
/// Every downgrade below is a session that would otherwise be attempted and never converge — the
/// cross-platform case most of all, where both devices have working UWB hardware and still cannot
/// range each other.
public enum ProximityPolicy {
    /// Returns a downgrade reason, or nil when ranging is worth attempting.
    public static func attempt(
        localSupportsUwb: Bool,
        peerAdvertisedUwb: Bool,
        peerToken: [UInt8]?,
        localIsAppleStack: Bool
    ) -> ProximityOutcome? {
        guard localSupportsUwb else { return .downgraded(why: "no-local-uwb") }
        guard peerAdvertisedUwb else { return .downgraded(why: "peer-not-uwb-capable") }
        guard let raw = peerToken else { return .downgraded(why: "no-peer-token") }
        guard let token = UwbTokenCodec.decode(raw) else { return .downgraded(why: "unreadable-peer-token") }

        let peerIsAppleStack: Bool
        if case .opaque = token { peerIsAppleStack = true } else { peerIsAppleStack = false }
        if peerIsAppleStack != localIsAppleStack {
            return .downgraded(why: "cross-platform-uwb-unsupported")
        }
        return nil
    }
}

File: kmp/src/commonMain/kotlin/tech/qrlesspay/sdk/Uwb.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) OpenBank contributors. Licensed under the Apache License, Version 2.0.
package tech.qrlesspay.sdk

/**
 * Enhanced proximity (wire spec §5) — optional, never required.
 *
 * **The spec describes UWB in prose and gives it no wire format.** It says the capability is
 * "negotiated, best-effort" without defining how a device announces it or how ranging parameters
 * cross between two phones. This file is therefore a **v1.1 proposal**, not an implementation of
 * something normative, and it is marked as such everywhere it surfaces. The bits it proposes:
 * advert flag `0x8` for UWB-capable, and a fourth GATT characteristic carrying the token below.
 *
 * Two facts decide the whole design, and both are the unwelcome kind:
 *
 * 1. **Apple's Nearby Interaction and Android's FiRa stack do not interoperate.** A cross-platform
 *    pair cannot range, at any effort, and the honest response is to detect that in the codec and
 *    downgrade rather than to attempt a session that will never converge.
 * 2. **UWB hardware is a minority** — iPhone 11 and later excluding the SE, and a slice of flagship
 *    Android. Most devices have no radio at all, so RSSI remains the baseline and UWB may only ever
 *    sharpen it.
 *
 * Consequence for the security story: UWB is the only *cryptographic* answer to a relay attack, and
 * it is unavailable on most pairs. Nothing here may become a precondition for paying.
 */
object Uwb {

    /** Advert flag proposed for "this device can range". Reserved-unused until the spec adopts it. */
    const val FLAG_UWB_CAPABLE = 0x8

    /** Proposed fourth characteristic, for exchanging the token below. */
    const val CHAR_UWB_TOKEN_UUID = "0000C3A6-2F3B-4E8A-9A5E-0B0E6F1C2D3A"
}

/**
 * A ranging token as it crosses the air.
 *
 * The platforms disagree about what a token even is. Android/FiRa needs concrete session parameters
 * and distinguishes the **controller** (which owns them) from the **controlee** (which joins);
 * Apple's `NIDiscoveryToken` is an opaque blob with no such split. The encoding below carries all
 * three shapes and, critically, tags them — so an Apple token arriving at an Android parser is
 * *rejected* rather than read as a malformed FiRa token and acted on.
 *
 * That rejection is the point. Silently misparsing a foreign token produces a ranging session that
 * reports distances derived from nonsense, which is far worse than no ranging at all: it would give
 * a payer a confident "0.3 m" for a peer that could be anywhere.
 */
sealed interface UwbToken {
    /** FiRa controller: owns the session parameters the controlee joins. */
    data class Controller(
        val address: ByteArray,
        val channel: Int,
        val preambleIndex: Int,
        val sessionId: Int,
    ) : UwbToken {
        override fun equals(other: Any?): Boolean = this === other ||
            (
                other is Controller && address.contentEquals(other.address) &&
                    channel == other.channel && preambleIndex == other.preambleIndex && sessionId == other.sessionId
                )

        override fun hashCode(): Int {
            var r = address.contentHashCode()
            r = 31 * r + channel
            r = 31 * r + preambleIndex
            r = 31 * r + sessionId
            return r
        }
    }

    /** FiRa controlee: only its address; the controller supplies everything else. */
    data class Controlee(val address: ByteArray) : UwbToken {
        override fun equals(other: Any?): Boolean = this === other ||
            (other is Controlee && address.contentEquals(other.address))

        override fun hashCode(): Int = address.contentHashCode()
    }

    /** Apple `NIDiscoveryToken`, opaque by design — never parsed, only handed back to the platform. */
    data class Opaque(val blob: ByteArray) : UwbToken {
        override fun equals(other: Any?): Boolean = this === other ||
            (other is Opaque && blob.contentEquals(other.blob))

        override fun hashCode(): Int = blob.contentHashCode()
    }
}

/**
 * `[magic 'Q'][version][kind][payload]`.
 *
 * The magic byte and version exist so that a token from the other platform, or from a future
 * revision, fails to decode instead of decoding into something wrong. The app this was extracted
 * from learned that the hard way and asserts it in a test; the same assertion lives in the
 * conformance corpus here, in both languages.
 */
object UwbTokenCodec {

    private const val MAGIC = 'Q'.code.toByte()
    private const val VERSION: Byte = 1
    private const val KIND_CONTROLLER: Byte = 1
    private const val KIND_CONTROLEE: Byte = 2
    private const val KIND_OPAQUE: Byte = 3
    private const val HEADER = 3
    private const val ADDRESS_BYTES = 2
    private const val CONTROLLER_PAYLOAD = ADDRESS_BYTES + 1 + 1 + 4
    private const val BYTE_MASK = 0xFF

    fun encode(token: UwbToken): ByteArray = when (token) {
        is UwbToken.Controller -> byteArrayOf(MAGIC, VERSION, KIND_CONTROLLER) +
            token.address.copyOf(ADDRESS_BYTES) +
            byteArrayOf(token.channel.toByte(), token.preambleIndex.toByte()) +
            ByteArray(4) { ((token.sessionId ushr (24 - it * 8)) and BYTE_MASK).toByte() }

        is UwbToken.Controlee -> byteArrayOf(MAGIC, VERSION, KIND_CONTROLEE) + token.address.copyOf(ADDRESS_BYTES)

        is UwbToken.Opaque -> byteArrayOf(MAGIC, VERSION, KIND_OPAQUE) + token.blob
    }

    /** Null for anything that is not a token of this version — including the other platform's. */
    fun decode(bytes: ByteArray): UwbToken? {
        if (bytes.size < HEADER) return null
        if (bytes[0] != MAGIC || bytes[1] != VERSION) return null
        val payload = bytes.copyOfRange(HEADER, bytes.size)
        return when (bytes[2]) {
            KIND_CONTROLLER -> {
                if (payload.size != CONTROLLER_PAYLOAD) return null
                UwbToken.Controller(
                    address = payload.copyOfRange(0, ADDRESS_BYTES),
                    channel = payload[ADDRESS_BYTES].toInt() and BYTE_MASK,
                    preambleIndex = payload[ADDRESS_BYTES + 1].toInt() and BYTE_MASK,
                    sessionId = (0 until 4).fold(0) { acc, i ->
                        (acc shl 8) or (payload[ADDRESS_BYTES + 2 + i].toInt() and BYTE_MASK)
                    },
                )
            }
            KIND_CONTROLEE -> if (payload.size != ADDRESS_BYTES) null else UwbToken.Controlee(payload)
            KIND_OPAQUE -> if (payload.isEmpty()) null else UwbToken.Opaque(payload)
            else -> null
        }
    }
}

/** What a ranging attempt concluded. */
sealed interface ProximityOutcome {
    /** UWB ranged and the peer is within [metres]. */
    data class Ranged(val metres: Double) : ProximityOutcome

    /**
     * No UWB session was possible — no radio on one side, a cross-platform pair, a rejected token,
     * or the user declined. **Not an error**: the RSSI baseline already gated this tile, and the
     * §6 confirmation is what authorises the payment either way.
     */
    data class Downgraded(val why: String) : ProximityOutcome
}

/**
 * Decides whether a ranging attempt is even worth starting, before any radio is touched.
 *
 * Every branch that returns [ProximityOutcome.Downgraded] here is a session that would otherwise be
 * attempted and never converge — the cross-platform case most of all, where both devices have
 * working UWB hardware and still cannot range each other.
 */
object ProximityPolicy {

    /** Peer capability, as read from the advert flag and the token it published. */
    fun attempt(
        localSupportsUwb: Boolean,
        peerAdvertisedUwb: Boolean,
        peerToken: ByteArray?,
        localIsAppleStack: Boolean,
    ): ProximityOutcome? {
        if (!localSupportsUwb) return ProximityOutcome.Downgraded("no-local-uwb")
        if (!peerAdvertisedUwb) return ProximityOutcome.Downgraded("peer-not-uwb-capable")
        val raw = peerToken ?: return ProximityOutcome.Downgraded("no-peer-token")
        val token = UwbTokenCodec.decode(raw) ?: return ProximityOutcome.Downgraded("unreadable-peer-token")

        val peerIsAppleStack = token is UwbToken.Opaque
        if (peerIsAppleStack != localIsAppleStack) {
            // Both sides have UWB and still cannot range: Apple's peer protocol is not FiRa. Detected
            // here so the UI falls straight back to the tap list instead of waiting out a session
            // that will never converge.
            return ProximityOutcome.Downgraded("cross-platform-uwb-unsupported")
        }
        return null // worth attempting; the platform ranger takes it from here
    }
}
